astacus/coordinator/plugins/clickhouse/steps.py

```diff
@@ -151,6 +153,28 @@ async def run_step(self, cluster: Cluster, context: StepsContext) -> Sequence[Ac
         return access_entities
 
 
+@dataclasses.dataclass
+class RetrieveUDFsStep(Step[Sequence[UDF]]):
+    zookeeper_client: ZooKeeperClient
+    replicated_user_defined_zookeeper_path: str | None
+
+    async def run_step(self, cluster: Cluster, context: StepsContext) -> Sequence[UDF]:
+        if self.replicated_user_defined_zookeeper_path is None:
+            return []
+
+        udfs: list[UDF] = []
+        async with self.zookeeper_client.connect() as connection:
+            change_watch = ChangeWatch()
+            children = await connection.get_children(self.replicated_user_defined_zookeeper_path, watch=change_watch)
+            for child in children:
+                udf_path = os.path.join(self.replicated_user_defined_zookeeper_path, child)
```

Comment:
nit. The other places where we handle ZooKeeper paths are intentionally not using `os.path` because ZooKeeper znode paths are not filesystem paths.
(On linux, it _should_ be fine, since I don't think it's possible to create a child node with '/' in its name, so just a nit because it's odd to have inconsistent styles in the same file, shouldn't be a security issue.)

Reply:
I think it is more about allowing for people to have a trailing slash on their zookeeper paths.